### ISSUE TITLE
No Mention of Slicing vs Paging. Is this really something we want to leave out of the spec? 

### ISSUE THREAD

**Issue by blima-ebsco** [EBSCOIS/api-standards#10]
Im just throwing this out there but in my past experiences allowing Paging at all in an API spec is trouble. If this is our "Perfect" spec we should be disallowing Paging explicitly and only allow Slicing.

I know this is a contentious issue to bring up and we may need exceptions because of legacy behavior. 

Thoughts?

An example below from the doc that favors Paging.
```
{
    "total_items": "166",
    "total_pages": "83",
    "users": [
    {
        "given_name": "James",
        "surname": "Greenwood",
        ...
        "links": [
            {
                "href": "https://api.foo.com/v1/customer/users/ALT-JFWXHGUV7VI",
                "rel": "self"
            }
        ]
    },
    {
        "given_name": "David",
        "surname": "Brown",
        ...
        "links": [
            {
                "href": "https://api.foo.com/v1/customer/users/ALT-MDFSKFGIFJ86DSF",
                "rel": "self"
            }
   },
}
```

**Comment by bmckinney-ebsco** [comment]
I think Elasticsearch provides a nice model, particularly when you're talking about the EDS API.

See: https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-from-size.html

**Comment by blima-ebsco** [comment]
@bmckinney-ebsco great example! Thats sort of what I am getting at, Pages are really in direct conflict with distributed systems. IE COUNT(*) is not only costly but almost always completely unnecessary. I pose we prohibit Paging out in general and stipulate Slicing. I can speak from a Spring point of view JPA already supports it out of the box.

**Comment by efrierson** [comment]
One of the benefits of our current page-oriented, state-aware EDS API is that you can take action on the current results.  For example, you could be 3 pages deep into a results set and then want to take the "Limit Results to Full Text" action.  You send your existing state, plus that action, and the API understands to:

   * Apply the new limiter AND
   * Respond with page 1, NOT with page 3 of your results (even though your current results are at page 3).  

In a slice environment, does that push the responsibility to the application to remember to also request the first slice, or is slicing/paging an independent issue that will have no impact on action-oriented API interaction?  I understand many popular OS applications like Blacklight and VuFind already know how to consume slice-oriented indexes like Elasticsearch and in fact prefer it, so if it is an either/or, I would probably lean slicing, even with losing the action-oriented syntax.

**Comment by esimons** [comment]
@blima-ebsco I'm honestly not familiar with how Slicing differs from Paging as presented via REST APIs, and a cursory Google search isn't giving me much (saturated with JPA-related results). Are you essentially talking about dropping the `total_items` and `total_pages` fields?

**Comment by blima-ebsco** [comment]
> @blima-ebsco I'm honestly not familiar with how Slicing differs from Paging as presented via REST APIs, and a cursory Google search isn't giving me much (saturated with JPA-related results). Are you essentially talking about dropping the `total_items` and `total_pages` fields?

@esimons Yes dropping `total_items` and `total_pages` fields is essentially the difference between Paging and Slicing.

**Comment by blima-ebsco** [comment]
> One of the benefits of our current page-oriented, state-aware EDS API is that you can take action on the current results. For example, you could be 3 pages deep into a results set and then want to take the "Limit Results to Full Text" action. You send your existing state, plus that action, and the API understands to:
> 
> * Apply the new limiter AND
> * Respond with page 1, NOT with page 3 of your results (even though your current results are at page 3).
> 
> In a slice environment, does that push the responsibility to the application to remember to also request the first slice, or is slicing/paging an independent issue that will have no impact on action-oriented API interaction? I understand many popular OS applications like Blacklight and VuFind already know how to consume slice-oriented indexes like Elasticsearch and in fact prefer it, so if it is an either/or, I would probably lean slicing, even with losing the action-oriented syntax.

@efrierson In the case you mentioned yes the application would have to remember it is supposed to request the first page when changing search criteria. In the past I  have made the `slice_index` default to 1 and set a reasonable default for 'num_items'. That gets you the same functionality, if I am understanding you correctly.

**Comment by esimons** [comment]
@blima-ebsco there's a section in the original `patterns` doc that doesn't exist in the current EBSCO fork, what do you think of their guidance re: the `total_required` and `page_token` params? If we carried this over to our docs, would it in any way help address your concerns?
https://github.com/paypal/api-standards/blob/master/patterns.md#query-parameters

**Comment by blima-ebsco** [comment]
@esimons I think my problem is more related to the use of `total_items` at the end of the day that is the field that makes the real difference between Paging and Slicing. I like the `page_token` idea if we can pull that off but I'm not sure we can require it.
The main reason to move to slicing is to get away from making expensive and almost always unnecessary `total_items` COUNT(*) call.

**Comment by esimons** [comment]
@blima-ebsco I just meant that that section of the Patterns doc seems to indicate that services which provide paging are not *required* to return either the `total_items` or `total_pages` in the response, and that even if they do provide the ability to return that data in the response, it should not be returned by default and instead must be explicitly requested by the caller. That seems like a reasonable policy to me.
It might be unnecessary for many use-cases, but for front-end pagination display it's kinda vital.
